### PR TITLE
Bug 1804258 - New SearchTermSuggestionsProvider

### PR DESCRIPTION
**Pull request by Mugurell**
The new provider will show by default up to 2 search suggestions based on past searches the user has done with the current search engine.

How the results would look in Fenix (ordering of the suggestions to be handled separately):

https://user-images.githubusercontent.com/11428869/208057932-eeaf22ac-f94a-4263-988b-e534b2988ed0.mp4

### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/main/docs/changelog.md) or does not need one
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/main/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1804258

### CODE REVIEW

**Comment by Mugurell** [comment]
Asking @Verdi if the result shown in the above video matches the expectations.
I want to confirm:
- the position of the new suggestions (below the action suggestion and the 3 search suggestions)
- the number of the new suggestions (2 at maximum)
- the search url shown below the suggestion.

**Comment by Verdi** [comment]
> Asking @Verdi if the result shown in the above video matches the expectations. I want to confirm:
> 
> * the position of the new suggestions (below the action suggestion and the 3 search suggestions)
> * the number of the new suggestions (2 at maximum)
> * the search url shown below the suggestion.

The number of suggestions (2) is good. But let's try to match desktop with the rest:
* search history suggestions above the search suggestion but below the action suggestion
* don't show the search url

**Comment by Mugurell** [comment]
Hardcoded in AC the position of the new suggestions between the action provider's ones and search provider's ones.
Updated the video from the PR description to show the result.

**Comment by Mugurell** [comment]
Small update needed.
Talked with Michael Verdi about this and while the initial specs asked to filter search terms by engine it would help more to not do this filtering and show the latest two search terms.
This would be useful in scenarios like doing a search with one particular search engine and then switch to another search engine and have the previous search terms suggested to redo the search with another engine.

File: android-components/components/browser/storage-sync/src/main/java/mozilla/components/browser/storage/sync/PlacesStorage.kt [hunk 1]
```diff
@@ -93,6 +99,23 @@ abstract class PlacesStorage(
...
+    /**
+     * Cleans up pending read operations of a specific query.
+     *
+     * @param queryToCancel Previous query to cancel reads for.
+     * Calling cancel multiple times for the same query has effect only the first time.
+     * Use this in scenarios where the same instance is used in multiple scenarios to prevent cases
+     * in which a general cancel operation for one scenario cancels other reads for the same query.
+     * If the value is an empty string all current reads are immediately cancelled.
+     */
+    override fun cancelReadsOfQuery(queryToCancel: String) {
```
Comment (gabrielluong) [inline android-components/components/browser/storage-sync/src/main/java/mozilla/components/browser/storage/sync/PlacesStorage.kt:111]: Driveby - Maybe we can rename this to `cancelQueryReads`?

**Review by gabrielluong — COMMENTED** [review]
Was just doing a quick driveby since I saw your standup note about needing a review and found this. Did you have a reviewer in mind? Perhaps @csadilek?

Otherwise, I can help take a look, but I am bit less familiar with SuggestionsProvider.

File: android-components/components/browser/storage-sync/src/main/java/mozilla/components/browser/storage/sync/PlacesStorage.kt (hunk 1 above)
Comment (rocketsroger) [reply android-components/components/browser/storage-sync/src/main/java/mozilla/components/browser/storage/sync/PlacesStorage.kt:111]: 🔍 nit: I think we can also just call this `cancelReads(nextQuery: String)`

File: android-components/components/browser/storage-sync/src/main/java/mozilla/components/browser/storage/sync/PlacesStorage.kt [hunk 2]
```diff
@@ -93,6 +99,23 @@ abstract class PlacesStorage(
...
+        if (queryToCancel.isEmpty() || lastCancelledQuery != queryToCancel) {
+            lastCancelledQuery = queryToCancel
```
Comment (rocketsroger) [inline android-components/components/browser/storage-sync/src/main/java/mozilla/components/browser/storage/sync/PlacesStorage.kt:113]: ❓ I thought `lastCancelledQuery` is not shared between storage? (ex: bookmarksStorage, historyStorage and historyMetadataStorage don't have the same `lastCancelledQuery`)

**Review by rocketsroger — COMMENTED** [review]
adding `SearchTermSuggestionsProvider` looks straightforward and I think it's ready to ship.  However, I have questions related to trying to prevent overlap cancels.

File: android-components/components/feature/awesomebar/src/main/java/mozilla/components/feature/awesomebar/provider/SearchTermSuggestionsProvider.kt [hunk 3]
```diff
@@ -22,6 +22,18 @@ import java.util.UUID
  */
 private const val DEFAULT_SUGGESTION_LIMIT = 2
 
+/**
+ * A too big limit but which help ensure the SearchSuggestionProvider' suggestions which should be placed
+ * below the ones from this provider will appear correctly.
+ */
+const val SEARCH_TERMS_MAXIMUM_ALLOWED_SUGGESTIONS_LIMIT: Int = 1_000
```
Comment (rocketsroger) [inline android-components/components/feature/awesomebar/src/main/java/mozilla/components/feature/awesomebar/provider/SearchTermSuggestionsProvider.kt:29]: ⚠️ Do you mean 1000?

Comment (rocketsroger) [inline android-components/components/feature/awesomebar/src/main/java/mozilla/components/feature/awesomebar/provider/SearchTermSuggestionsProvider.kt:29]: 🔍 nit: I think 1000 is readable here.  No need to underline.

File: android-components/components/browser/storage-sync/src/main/java/mozilla/components/browser/storage/sync/PlacesStorage.kt (hunk 2 above)
Comment (Mugurell) [reply android-components/components/browser/storage-sync/src/main/java/mozilla/components/browser/storage/sync/PlacesStorage.kt:113]: Good catch, tested now again.
Indeed Fenix is currently using two different instances of `PlacesStorage`, one for [bookmarks](https://searchfox.org/mozilla-mobile/rev/7c70c3df303e19acc69877821c3770fe85393005/fenix/app/src/main/java/org/mozilla/fenix/components/Core.kt#357) and one for [history](https://searchfox.org/mozilla-mobile/rev/7c70c3df303e19acc69877821c3770fe85393005/fenix/app/src/main/java/org/mozilla/fenix/components/Core.kt#356).
So between the providers of this data there should be no overlapping.

But on Fenix the `CombinedHistoryProvider` will use [the same `PlacesStorage` instance](https://searchfox.org/mozilla-mobile/rev/7c70c3df303e19acc69877821c3770fe85393005/fenix/app/src/main/java/org/mozilla/fenix/search/awesomebar/AwesomeBarView.kt#126-127) for both metadata and "legacy" history.
And then I planned to use this new provider with the same instance also - as the instance is quite big and we're interested in only a very small part of it's functionality.
This will mean that we need to somehow prevent overlapping cancel requests which is why I've added this new `cancelReads(nextQuery: String)` api.
(Using the same storage instance for multiple queries is no less performant than using different storage instances for multiple queries as all calls will query the same database from application-services so they would be executed serially either way.)
So I found it simpler, probably better for performance - memory and safer for scenarios in which the same underlying storage is used for multiple providers - prevent unintentional overlapping cancel requests.

**Review by rocketsroger — APPROVED** [review]

**Comment by Mugurell** [comment]
Thank you Roger for the help!
Not sure though what the problem is in the [`lint-buildconfig-android-components`](https://github.com/mozilla-mobile/firefox-android/pull/263/checks?check_run_id=10907394729) task, never seen this before:
> [task 2023-01-26T16:27:39.228Z] + taskcluster/scripts/lint/is_buildconfig_yml_up_to_date.py android-components
[task 2023-01-26T16:27:39.483Z] 2023-01-26 16:27:39,482 - INFO - Looking for dependencies in api configuration in android-components
[task 2023-01-26T16:29:44.345Z] 2023-01-26 16:29:44,345 - INFO - Looking for dependencies in compileOnly configuration in android-components
[task 2023-01-26T16:30:21.728Z] 2023-01-26 16:30:21,728 - INFO - Looking for dependencies in implementation configuration in android-components
[task 2023-01-26T16:30:58.189Z] 2023-01-26 16:30:58,189 - INFO - Looking for dependencies in testImplementation configuration in android-components
[task 2023-01-26T16:31:33.554Z] 2023-01-26 16:31:33,553 - INFO - Updated android-components/.buildconfig.yml with latest gradle config!
[task 2023-01-26T16:31:33.732Z] 2023-01-26 16:31:33,732 - ERROR - .buildconfig.yml file changed! Please update it by running:
[task 2023-01-26T16:31:33.733Z] 
[task 2023-01-26T16:31:33.733Z] curl --location https://firefox-ci-tc.services.mozilla.com/api/queue/v1/task/NLDNc8tTR-akuZ0gWTz5pQ/artifacts/public%2Fgit%2Fbuildconfig.diff | gunzip | git apply
[task 2023-01-26T16:31:33.733Z] 
[task 2023-01-26T16:31:33.733Z] Then commit and push!

**Comment by csadilek** [comment]
> [task 2023-01-26T16:31:33.732Z] 2023-01-26 16:31:33,732 - ERROR - .buildconfig.yml file changed! Please update it by running:
[task 2023-01-26T16:31:33.733Z]
[task 2023-01-26T16:31:33.733Z] curl --location https://firefox-ci-tc.services.mozilla.com/api/queue/v1/task/NLDNc8tTR-akuZ0gWTz5pQ/artifacts/public%2Fgit%2Fbuildconfig.diff | gunzip | git apply
[task 2023-01-26T16:31:33.733Z]
[task 2023-01-26T16:31:33.733Z] Then commit and push!

@JohanLorenzo It is working! 🎉  :)

@Mugurell This is because this patch introduced a new dependency here https://github.com/mozilla-mobile/firefox-android/pull/263/files#diff-4230cf662f55c5246dedb219c9c518c176d177f50394bfcb9a1cb012c59639bbR32 and we're caching the dependency graph in .buildconfig.yml for faster builds!
